compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/diagnotic/BDiagnostic.java

```diff
@@ -79,10 +79,33 @@ public String getMessage() {
     public DiagnosticCode getCode() {
         return code;
     }
-    
+
+    @Override
+    public int hashCode() {
+        return kind.hashCode() + pos.hashCode() + msg.hashCode() + code.hashCode();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (!(obj instanceof BDiagnostic)) {
```

Comment:
One of the best practice is to check reference equality first. Shall we add that?

Reply:
Resolved with https://github.com/ballerina-platform/ballerina-lang/pull/18233/commits/eaa8b93163c4386f1dd358a359e7a8ec655d8e4a.